packages/telemetry/src/telemetryReporter.ts

```diff
@@ -55,21 +68,75 @@ export class TelemetryReporter extends AsyncCreatable<TelemetryOptions> {
    * @param attributes {Attributes} - map of properties to publish alongside the event.
    */
   public sendTelemetryEvent(eventName: string, attributes: Attributes = {}): void {
+    const name = `${this.options.project}/${eventName}`;
+    const { properties, measurements } = buildPropertiesAndMeasurements(attributes);
+    this.sendTelemetry(TelemetryMethod.EVENT, name, { name, properties, measurements });
+  }
+
+  /**
+   * Publishes exception to app insights dashboard
+   * @param exception {Error} - exception you want published.
+   * @param attributes {Attributes} - map of measurements to publish alongside the exception.
+   */
+  public sendTelemetryException(exception: Error, attributes: Attributes = {}): void {
+    const { properties, measurements } = buildPropertiesAndMeasurements(attributes);
+    this.sendTelemetry(TelemetryMethod.EXCEPTION, exception.message, { exception, properties, measurements });
+  }
+
+  /**
+   * Publishes diagnostic information to app insights dashboard
+   * @param message {string} - trace message to sen to app insights.
+   * @param properties {Properties} - map of properties to publish alongside the event.
+   */
+  public sendTelemetryTrace(traceMessage: string, properties?: Properties): void {
+    this.sendTelemetry(TelemetryMethod.TRACE, traceMessage, { message: traceMessage, properties });
+  }
+
+  /**
+   * Publishes metric to app insights dashboard
+   * @param name {string} - name of the metric you want published
+   * @param value {number} - value of the metric
+   * @param properties {Properties} - map of properties to publish alongside the event.
+   */
+  public sendTelemetryMetric(metricName: string, value: number, properties?: Properties): void {
+    this.sendTelemetry(TelemetryMethod.METRIC, metricName, { name: metricName, value, properties });
+  }
+
+  private sendTelemetry(
+    method: TelemetryMethod,
+    message: string,
+    data: EventTelemetry | ExceptionTelemetry | MetricTelemetry | TraceTelemetry
+  ): void {
     if (!isSfdxTelemetryEnabled(this.env)) return;
 
     if (this.appInsightsClient) {
-      const name = `${this.options.project}/${eventName}`;
-      const { properties, measurements } = buildPropertiesAndMeasurements(attributes);
-      this.logger.debug(`Sending telemetry event: ${name}`);
+      this.logger.debug(`Sending telemetry: ${message}`);
       try {
-        this.appInsightsClient.trackEvent({ name, properties, measurements });
+        switch (method) {
+          case TelemetryMethod.EVENT: {
+            this.appInsightsClient.trackEvent(data as EventTelemetry);
+            break;
+          }
+          case TelemetryMethod.EXCEPTION: {
+            this.appInsightsClient.trackException(data as ExceptionTelemetry);
```

Comment:
One thing we had to do on VSCode's side was to override the stack value when logging exceptions since the appinsights module adds it out-of-the-box and it will contain sensitive user info that violates GDPR. It's worth checking to see if it will be needed in this implementation as well.

Reply:
@jgellin-sf I understand that won't be an issue for Web IDE scenario but this is a shared library that is used by the CLI and other CLI plugins. The moment they start using these changes they will be creating GDPR violations. I think it's better if that is addressed right now or this could slip and become a problem.

Reply:
@lcampos Thomas merged this PR just as you were responding. I captured this discussion in GUS here for follow-up: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000007rnlfIAA/view